### PR TITLE
Deprecated usecols with out of bounds indices in read_csv

### PR DESCRIPTION
**Pull request by phofl**
- [x] xref #25623
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

This currently raises 

<details>
Traceback (most recent call last):
  File "/home/developer/.config/JetBrains/PyCharm2021.1/scratches/scratch_4.py", line 539, in <module>
    pd.read_csv('test.csv', header=0, usecols=[0, 10], engine='python')
  File "/home/developer/PycharmProjects/pandas/pandas/io/parsers/readers.py", line 552, in read_csv
    return _read(filepath_or_buffer, kwds)
  File "/home/developer/PycharmProjects/pandas/pandas/io/parsers/readers.py", line 471, in _read
    return parser.read(nrows)
  File "/home/developer/PycharmProjects/pandas/pandas/io/parsers/readers.py", line 998, in read
    index, columns, col_dict = self._engine.read(nrows)
  File "/home/developer/PycharmProjects/pandas/pandas/io/parsers/python_parser.py", line 285, in read
    data, columns = self._exclude_implicit_index(alldata)
  File "/home/developer/PycharmProjects/pandas/pandas/io/parsers/python_parser.py", line 303, in _exclude_implicit_index
    names = [names[i] for i in sorted(self._col_indices)]
  File "/home/developer/PycharmProjects/pandas/pandas/io/parsers/python_parser.py", line 303, in <listcomp>
    names = [names[i] for i in sorted(self._col_indices)]
</details>
IndexError: list index out of range

on master but unfortunately works on 1.2.4, so we can either raise ParserError with 1.2.4 or fix and deprecate then to remove in 2.0, related to #41129

I think fixing and deprecating would be more sensible, but only realised that this works on 1.2.4 after finishing this, so wanted to put up for discussion at least :)

cc @gfyoung

### CODE REVIEW

File: doc/source/whatsnew/v1.3.0.rst [hunk 1]
```diff
@@ -797,6 +797,7 @@ I/O
 - Bug in :func:`read_excel` raising ``AttributeError`` with ``MultiIndex`` header followed by two empty rows and no index, and bug affecting :func:`read_excel`, :func:`read_csv`, :func:`read_table`, :func:`read_fwf`, and :func:`read_clipboard` where one blank row after a ``MultiIndex`` header with no index would be dropped (:issue:`40442`)
 - Bug in :meth:`DataFrame.to_string` misplacing the truncation column when ``index=False`` (:issue:`40907`)
 - Bug in :func:`read_orc` always raising ``AttributeError`` (:issue:`40918`)
+- Bug in :func:`read_csv` raising uncontrolled ``ValueError`` when ``usecols`` index is ouf of bounds, now raising ``ParserError`` (:issue:`25623`)
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v1.3.0.rst:800]: im not sure what "uncontrolled" means here

Comment (phofl) [reply doc/source/whatsnew/v1.3.0.rst:800]: Not raised on purpose by us but instead raised because we are accessing a non existent list index

**Comment by jreback** [comment]
> unfortunately works on 1.2.4,

what does this mean?

**Comment by phofl** [comment]
This is a regression on master compared to 1.2.x series. So we should probably fix and then deprecate to not change behavior in 1.3

Unfortunately means, if this would not have worked on 1.2.x we could immediately start raising a ParserError without worrying about backwarts compatibility

File: pandas/tests/io/parser/test_python_parser_only.py [hunk 2]
```diff
@@ -312,3 +312,19 @@ def test_malformed_skipfooter(python_parser_only):
     msg = "Expected 3 fields in line 4, saw 5"
     with pytest.raises(ParserError, match=msg):
         parser.read_csv(StringIO(data), header=1, comment="#", skipfooter=1)
+
+
+@pytest.mark.parametrize("header", [0, None])
+@pytest.mark.parametrize("names", [None, ["a", "b"], ["a", "b", "c"]])
+def test_usecols_indices_out_of_bounds(python_parser_only, names, header):
```
Comment (gfyoung) [inline pandas/tests/io/parser/test_python_parser_only.py:319]: Can this be tested with the `CParser` too?

File: pandas/io/parsers/python_parser.py [hunk 3]
```diff
@@ -506,13 +510,13 @@ def _infer_columns(self):
...
                     # Ignore output but set used columns.
-                    self._handle_usecols([names], names)
+                    self._handle_usecols([names], names, ncols)
                     columns = [names]
                     num_original_columns = ncols
 
         return columns, num_original_columns, unnamed_cols
 
-    def _handle_usecols(self, columns, usecols_key):
+    def _handle_usecols(self, columns, usecols_key, num_original_columns):
```
Comment (gfyoung) [inline pandas/io/parsers/python_parser.py:519]: Brief docstring on this new parameter to explain how it differs from `columns` (and why we couldn't just use `columns.length` in the logic).

File: pandas/tests/io/parser/test_python_parser_only.py (hunk 2 above)
Comment (phofl) [reply pandas/tests/io/parser/test_python_parser_only.py:319]: See #41129

**Comment by phofl** [comment]
@gfyoung was not sure, because this is working on 1.2.x without raising an error, it is simply ignoring the indexes out of range, but I would be fine with doin this immediately

**Comment by jreback** [comment]
> @gfyoung was not sure, because this is working on 1.2.x without raising an error, it is simply ignoring the indexes out of range, but I would be fine with doin this immediately

oh maybe i misuderstood. so if this was working on 1.2.x then we should deprecate first

**Comment by gfyoung** [comment]
> because this is working on 1.2.x without raising an error, it is simply ignoring the indexes out of range, but I would be fine with doin this immediately

Oh, I see!  The w/o raising an error part puts me in agree with @jreback then.

**Comment by phofl** [comment]
Yeah same on my side. Will mark this as draft until I have fixed the error on master. Then we can switch ParserError with FutureWarning

**Comment by phofl** [comment]
After #41244 was merged we can deprecate now for both engines

File: pandas/io/parsers/python_parser.py (hunk 3 above)
Comment (jreback) [reply pandas/io/parsers/python_parser.py:532]: can you type args here

**Review by jreback — CHANGES_REQUESTED** [review]
small request

**Comment by phofl** [comment]
Done. should we use from __future__ import annotations in a follow up?

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
> Done. should we use from **future** import annotations in a follow up?

yes for sure, as that's being done elsewhere in the codebase.


thanks @phofl

**Pull request merged**